### ISSUE TITLE
(feature request) timeout for pandas.io.json.read_json

### ISSUE THREAD

**Issue by altsheets** [pandas-dev/pandas#14280]
Please allow to set a `timeout` in (all) your internet requests. We have just identified a halting problem in a repeating long runner when using `read_json()`.

Example:

```
pandas.io.json.read_json(url, timeout=10.0)
```

My current workaround is to temporarily set `socket.setdefaulttimeout(10.0)` globally (default is `None`!). Which works, but if your read_json had already allowed to set it locally, I would have probably realized much earlier.

Thanks a lot for considering this. Thanks for pandas, an amazing tool.

**Comment by altsheets** [comment]
My workaround:

```
import socket
from pandas.io.json import read_json

def read_json_with_timeout(url, timeout):
    """
    pandas.io.json.read_json had no timeout option. This fixes that.
    """
    timeout_before = socket.getdefaulttimeout()
    socket.setdefaulttimeout(timeout)
    df = read_json(url, orient="columns")
    socket.setdefaulttimeout(timeout_before)
    
    return df
```

but better would be to give a timeout option everywhere where you query the web.

**Comment by bkandel** [comment]
You could also try using the `retrying` package, which provides decorators for retrying function calls that may fail (like internet queries).

**Comment by altsheets** [comment]
Thx. I had already been catching exceptions, and then simply retried, after a shorter-than-ususal sleep.

However, with the default `socket.getdefaulttimeout() --> None`, we ran into non-returning calls; actually very infrequent, sometimes only once per month. But then in those cases the whole backend halted without recovery, and at first me had no clue why.

Now, with `socket.setdefaulttimeout(10.0)` this should not happen anymore. Instead the `read_json' will throw an exception, and I can treat that like any other internet connection failure.

**Comment by mroeschke** [comment]
IMO this is out of scope for pandas. A few IO functions can read from urls but they are essentially treated like file handles. Wrapping pandas IO routines to handle timeout issues would be a better solution. Closing as there seems not to be much traction.
